### ISSUE TITLE
Duplicate host shows up when the host is restarted

### ISSUE THREAD

**Issue by cglewis** [rancher/rancher#936]
Here's an example showing the behavior:

![screen shot 2015-05-27 at 11 09 35 am](https://cloud.githubusercontent.com/assets/2931079/7843798/ffb0243a-0460-11e5-8fd4-608f153a7de4.png)

To reproduce, I just restarted the boot2docker host that was running the agent container, which automatically started the agent container again when it booted back up.  Afterwards, Rancher showed two copies of the host, which were also inconsistent.  Deactivating and deleting the correct (old) one made it go away, but this should be handled in a cleaner way for the end user, IMO.

**Comment by vincent99** [comment]
Rancher saves some state like the host's UUID in `/var/lib/rancher/` on the host.  But that part of boot2docker's disk is ephemeral by default, so when you reboot it's gone, gets a new UUID when the agent starts up, and is a "different" host.

Separately, the old one doesn't show disconnected because https://github.com/rancherio/cattle/issues/265 .

**Comment by Zaephor** [comment]
Experiencing this when deployed to a dedicated tower running RancherOS 0.3.1 as well.

Only command run on the system post ranchero-install `sudo docker run -d --privileged -v /var/run/docker.sock:/var/run/docker.sock rancher/agent:v0.7.8 http://other.systemrunning.server:8080/v1/scripts/generatedcode`

Then rebooting, causes the system to duplicate in rancher/server. Not sure if necessary or relevant, I have rancher/server running on a different dedicated tower from the one running rancher/agent.

**Comment by vincent99** [comment]
That should be the same issue, `/var/lib/rancher/` is ephemeral and gone on reboot.  We should probably work better on our own OS :smile:

**Comment by Zaephor** [comment]
lol I had a feeling that was the case, but thought I'd mention it with as much information as I can just to be sure :bowtie:

**Comment by imikushin** [comment]
I'm going to look what I can do to fix this in RancherOS

**Comment by pwFoo** [comment]
Similar problems with RancherOS und Rancher. After reboot connection seems broken. State is processing / stopping or starting forever.

**Comment by funkytaco** [comment]
Is this supposed to be fixed? Because it just happened to me:

time="2015-10-29T00:35:59Z" level=error msg="could not find host uuid" 
time="2015-10-29T00:35:59Z" level=error msg="could not find host uuid" 
time="2015-10-29T00:36:05Z" level=info msg="Registering backend for host [b60ad4d0-1a8b-4a21-b160-6240db776cf8]" 
time="2015-10-29T00:36:05Z" level=info msg="Creating Machine" eventId=7dc2883a-5d02-43dc-8fd2-d7d1d2dd1468 resourceId=1ph5 
time="2015-10-29T00:36:05Z" level=info msg="Activating Machine" eventId="802ad247-b9c4-4b78-a2eb-a58f4dd2612c" resourceId=1ph4 


Component	Version
Rancher	v0.42.0
Cattle	v0.102.0
User Interface	v0.57.0
Rancher Compose	v0.4.3

**Comment by imikushin** [comment]
@funkytaco what RancherOS version is your agent running on?

**Comment by funkytaco** [comment]
I'm not running RancherOS right now. I'm running the default docker machine. Did you want me to try and recreate the issue?

**Comment by Sleashe** [comment]
I'm experiencing the same issues. Running on mac os x using docker-toolbox. I tried to stop VMs in different orders but nothing changes. Rancher UI tells to add a host using this command line 

docker run -d --privileged -v /var/run/docker.sock:/var/run/docker.sock rancher/agent:v0.8.2 http://192.168.99.100:8080/v1/scripts/7C6B9C9DF88C7C817794:1446300000000:iDbJVsCfbjvnMAp4rf8MYMNJFc

Is there any way to mount /var/lib/rancher at container startup to keep UUID data persistent and avoid duplicate hosts ?

**Comment by imikushin** [comment]
@funkytaco @Sleashe Rancher (the container management platform) expects docker daemon to run on Linux x86_64 hosts with a persistent filesystem (so it can save state across reboots). 

boot2docker (a tiny Linux distro used by default by docker-machine) apparently can't persist `/var/lib/rancher` dir, and that's why you're having this problem. b2d is also explicitly aimed at development, and that means lots of short lived containers, and its sole purpose is running the docker daemon and that's why it's probably the best way to start with Docker on OS X. Using it for production environments is not recommended by its developers. 

In short, to run Rancher, use RancherOS instead of b2d on docker-machine, :
```
docker-machine create --driver virtualbox \  
  --virtualbox-boot2docker-url https://github.com/rancher/os/releases/download/v0.4.0/rancheros.iso \                           
  --virtualbox-memory "2048" ros-040
```

I'm going to close this issue now, but feel free to comment if you have any feedback or questions.

**Comment by Sleashe** [comment]
@imikushin Thanks for taking time to answer. Yes in fact, it's working well using RancherOs, but I can't mount volumes properly using RancherOs for development puposes. When using Virtual box driver when creating VM using docker-machine command, mounting volumes inside container is working well, but when using RancherOs I can't figure out how to solve this.

**Comment by deniseschannon** [comment]
@Sleashe Could you open a separate issue in https://github.com/rancher/os/issues about your questions on mounting volumes inside a container?

**Comment by edqallen** [comment]
This issue still occurs when using a dedicated rancher install (container on ubuntu) managing vsphere hosts, which use boot2docker.  If the host VMs are rebooted for any reason, duplicates show up in rancher's infrastructure tab.  The duplicates do not show any containers.

**Comment by aemneina** [comment]
@edqallen as a workaround you might want to have a volume attached to the rancher guest with the label RANCHER_STATE.

**Comment by edqallen** [comment]
@aemneina is this something that can be done from within the rancher "Add Host" advanced options with the vmwarevsphere driver?  These VMs are created by rancher with 20G of persistent storage, why isn't that storage being used for this?

**Comment by aemneina** [comment]
Sending down a volume label doesn't look like an option available in the docker-machine vsphere driver in use: https://docs.docker.com/machine/drivers/vsphere/

**Comment by edqallen** [comment]
@imikushin pointed out the real "issue" here, which I agree with, the question then becomes why is Rancher creating the VMs with b2d in the first place?  Why isn't it using rancherOS or some other prebuilt ISO image?  There *is* an option to specify the ISO that is used in the add host form, maybe the default rancherOS ISO will work if specified here.

**Comment by funkytaco** [comment]
Now that Docker for Mac & Windows are stable, can we use Rancher on them or will this issue come up?

**Comment by deniseschannon** [comment]
@funkytaco We still haven't declared support for Docker for Mac or Windows.

**Comment by vincent99** [comment]
The newer agent image detects boot2docker and will write the state somewhere that is persistent instead of ephemeral, so it shows up as the same host on reboot.  But supporting Docker for Mac/Windows is an entirely different problem because of all the hackery they do to try and pretend it's not running in a VM.  https://github.com/rancher/10acre-ranch/blob/master/bin/mac-ranch worked at one point during the beta but I think there's even less exposure of the VM now so there's no easy way to register a host to a server container running inside of it.  But I haven't had time to look lately.

**Comment by funkytaco** [comment]
@deniseschannon  Okay, thanks!

But then, maybe your tagline should be "A Platform for Operating Docker for Linux in Production http://rancher.com".  You could always change it back when you do support all three, which I'm sure isn't easy to do.

**Comment by vincent99** [comment]
It might be "released", but "stable" and "in production" are not words often used to describe Docker for Mac.  And it still _is_ Linux.

**Comment by mrlesmithjr** [comment]
I also just experienced this same exact thing. I built 5 Docker hosts using docker-machine on HyperV using BootToDocker.iso as the RancherOS would not work with Hyper-V driver. However, after rebooting all of the hosts they all show up duplicated and active.

**Comment by vincent99** [comment]
@mrlesmithjr the agent change would only be in 1.2.0-preX

**Comment by mrlesmithjr** [comment]
Ok good to know...Thanks! Also is there anyway to get RancherOS working with Hyper-V correctly? Would rather use that if I could.

**Comment by ryancorvetti** [comment]
For those still struggling to address this issue with rancher/agent running in boot2docker, run the following *before* spinning up the rancher/agent container:
```bash
docker-machine ssh [MACHINE_NAME] -- '
    sudo mkdir -p /mnt/sda1/var/lib/rancher \
    && sudo mkdir /var/lib/rancher \
    && echo "/mnt/sda1/var/lib/rancher /var/lib/rancher none defaults,bind 0 0" | sudo tee -a /etc/fstab 1>/dev/null \
    && sudo mount --bind /mnt/sda1/var/lib/rancher /var/lib/rancher
```
This will setup a bind mount to map /var/lib/rancher to /mnt/sda1/var/lib/rancher, the latter of which will persist between reboots of the docker machine.

**Comment by ryancorvetti** [comment]
For those using docker-machine (aka. boot2docker) with rancher...

If you are are interested in getting rancher/agent state to persist between reboots (and therefore not duplicate host entries in rancher/server), try the following *before* running the rancher/agent container on the target docker machine:

```bash
docker-machine ssh [DOCKER_MACHINE] -- '
        sudo mkdir -p /mnt/sda1/var/lib/rancher \
        && sudo mkdir /var/lib/rancher \
        && sudo mount --bind /mnt/sda1/var/lib/rancher /var/lib/rancher \
        && echo -e "#!/bin/sh\nmkdir /var/lib/rancher\nmount --bind /mnt/sda1/var/lib/rancher /var/lib/rancher" \
         | sudo tee -a /var/lib/boot2docker/bootlocal.sh 1>/dev/null \
        && sudo chmod +x /var/lib/boot2docker/bootlocal.sh
'
```

The above establishes a bind mount to map /var/lib/rancher to /mnt/sda1/var/lib/rancher, the latter of which will survive between reboots.

**Note:** Changing volume path using -v switch during `docker run` does *not* work, since /var/lib/rancher is not a data volume, but rather a parameter that is used by the run.sh entrypoint to setup the rancher/cattle environment.

**Comment by swartel** [comment]
+1 I did the same kind of operation in /var/lib/boot2docker/bootlocal.sh using rsync to restore /var/lib/rancher from a shared folder. And since no more duplicate host (boot2loader) after reboot.
